### PR TITLE
feat: add styling for forms

### PR DESCRIPTION
**Pull request by inikolova**
Related to https://github.com/telerik/kendo-theme-default/pull/391

Closes: https://github.com/telerik/k2-site/issues/155

### CODE REVIEW

**Review by gyoshev — APPROVED** [review]

File: scss/input/_layout.scss [hunk 1]
```diff
@@ -5,13 +5,16 @@
     $textbox-shadow: 0 2px 2px 1px rgba(0, 0, 0, .06) !default;
     $fieldset-margin: 30px;
 
+    $input-border-width: 1px !default;
+
     .k-textbox,
-    .k-input.k-textbox {
+    .k-input.k-textbox,
+    .k-textarea {
```
Comment (joneff) [inline scss/input/_layout.scss]: Do we need a separate class for textarea?

File: scss/input/_layout.scss [hunk 2]
```diff
@@ -22,6 +25,11 @@
         -webkit-appearance: none;
     }
 
+    .k-textarea {
```
Comment (joneff) [inline scss/input/_layout.scss:26]: To continue the the above, this line could be

```scss
textarea.k-textbox {
    // styles go here
}
```

File: scss/input/_layout.scss [hunk 3]
```diff
@@ -77,7 +85,7 @@
     .k-checkbox-label,
     .k-radio-label {
         position: relative;
-        padding-left: $spacer-x * 1.5;
+        padding-left: $icon-size + ($padding-x-lg / 2);
```
Comment (joneff) [inline scss/input/_layout.scss]: this will yield an error in theme-bootstrap

File: scss/input/_layout.scss [hunk 4]
```diff
@@ -191,6 +199,122 @@
...
+
+            &:first-child:first-of-type {
+                margin-top: 0;
+            }
+
+            &:last-child:last-of-type {
+                margin-bottom: 0;
+            }
+
+            legend {
```
Comment (joneff) [inline scss/input/_layout.scss]: the legend should always be in a fieldset, so we can "flatten" this selector

File: scss/input/_layout.scss [hunk 5]
```diff
@@ -191,6 +199,122 @@
...
+                font-weight: 600;
+                line-height: 1;
+                margin-bottom: $padding-y-lg * 2;
+                text-transform: uppercase;
+                padding: 0 $padding-x 0 0;
+                width: auto;
+            }
+        }
+
+        .k-form-field {
```
Comment (joneff) [inline scss/input/_layout.scss]: We can flatten this selector too

File: scss/input/_layout.scss [hunk 6]
```diff
@@ -191,6 +199,122 @@
...
+                margin-bottom: 0;
+            }
+        }
+
+        .k-alert-error {
+            font-size: $font-size-sm;
+            margin-top: $padding-y * 2;
+        }
+
+        .k-checkbox-label,
```
Comment (joneff) [inline scss/input/_layout.scss]: This could be merged with the above standalone styles

File: scss/input/_layout.scss [hunk 7]
```diff
@@ -191,6 +199,122 @@
...
+                display: block;
+                padding: $padding-y 0;
+            }
+
+            > input {
+                width: 100%;
+            }
+        }
+
+        .k-field-info {
```
Comment (joneff) [inline scss/input/_layout.scss]: what if I have inline form with field info?

File: scss/input/_layout.scss [hunk 8]
```diff
@@ -191,6 +199,122 @@
...
+                box-sizing: border-box;
+            }
+
+            > span,
+            > label:not(.k-checkbox-label):not(.k-radio-label) {
+                width: $inline-form-element-width;
+                float: left;
+                text-align: right;
+                line-height: $line-height;
+                padding: ($padding-y + $input-border-width) 0;
```
Comment (joneff) [inline scss/input/_layout.scss]: this will yield error in theme-bootstrap

File: scss/input/_layout.scss [hunk 9]
```diff
@@ -191,6 +199,122 @@
         }
     }
 
+    // Forms
+    $form-spacer: $padding-x * 2 !default;
```
Comment (joneff) [inline scss/input/_layout.scss:201]: can we either use `$spacer`, which is meant for margins, or a new variable?

File: scss/input/_theme.scss [hunk 10]
```diff
@@ -42,6 +43,10 @@
             background-color: $accent;
             color: $accent-contrast;
         }
+
+        &::placeholder {
+            color: rgba($text-color, .5);
```
Comment (joneff) [inline scss/input/_theme.scss]: this should be `rgba($input-text, .5)`

File: scss/input/_layout.scss [hunk 11]
```diff
@@ -5,13 +5,16 @@
     $textbox-shadow: 0 2px 2px 1px rgba(0, 0, 0, .06) !default;
     $fieldset-margin: 30px;
 
+    $input-border-width: 1px !default;
```
Comment (joneff) [inline scss/input/_layout.scss]: we can extract this variable to _variables.scss

File: scss/input/_layout.scss [hunk 12]
```diff
@@ -191,6 +199,122 @@
...
+    // Forms
+    $form-spacer: $padding-x * 2 !default;
+
+    .k-form,
+    .k-form-inline {
+        font-size: $font-size;
+        line-height: $line-height;
+        padding: $form-spacer;
+
+        fieldset {
```
Comment (joneff) [inline scss/input/_layout.scss:209]: shouldn't `.k-form fieldset` and `.k-fieldset` have the same styles?

File: scss/input/_layout.scss [hunk 13]
```diff
@@ -191,6 +199,122 @@
...
+            display: inline-block;
+            font-size: ($font-size-xs / $font-size) * 1em;
+            margin: 0 $padding-x;
+        }
+    }
+
+    .k-form-inline {
+        $inline-form-element-width: 25% !default;
+
+        .k-form-field {
```
Comment (joneff) [inline scss/input/_layout.scss:286]: Are there any arguments against `display: flex`?

File: scss/input/_layout.scss [hunk 14]
```diff
@@ -191,6 +199,122 @@
...
+
+    .k-form {
+        input,
+        label:not(.k-checkbox-label):not(.k-radio-label),
+        .k-widget {
+            display: block;
+        }
+
+        .k-form-field {
+            > span {
```
Comment (joneff) [inline scss/input/_layout.scss]: What happens if I have a combobox as the input? It renders a span.

File: scss/input/_layout.scss [hunk 15]
```diff
@@ -191,6 +199,122 @@
...
+        }
+    }
+
+    .k-form-inline {
+        $inline-form-element-width: 25% !default;
+
+        .k-form-field {
+            > span,
+            > input {
+                display: inline-block;
```
Comment (joneff) [inline scss/input/_layout.scss]: do we really need this selector? Floated elements can have dimensions. The `box-sizing` property is defined in the next `span` selector, so it only needs to exist in the next `input` selector.

File: scss/input/_layout.scss [hunk 16]
```diff
@@ -191,6 +199,122 @@
...
+                box-sizing: border-box;
+            }
+
+            > span,
```
Comment (joneff) [inline scss/input/_layout.scss]: Same question as with block forms -- what if my input is combobox? It renders a span.

File: scss/input/_layout.scss (hunk 12 above)
Comment (inikolova) [reply scss/input/_layout.scss:209]: It appears that **.k-fieldset** has been previously defined, but not documented. **.k-form fieldset** follows the idea for not using additional classes for styling form elements

File: scss/input/_layout.scss (hunk 9 above)
Comment (inikolova) [reply scss/input/_layout.scss:201]: $padding-x / $padding-y are used for margins in other places, too. I don't think we need additional variable here

File: scss/input/_layout.scss (hunk 1 above)
Comment (inikolova) [reply scss/input/_layout.scss]: I think it is more meaningful to use **.k-textarea** for textarea elements. So I prefer keeping this class

File: scss/input/_layout.scss (hunk 2 above)
Comment (inikolova) [reply scss/input/_layout.scss:26]: Same as the above

File: scss/input/_layout.scss (hunk 12 above)
Comment (gyoshev) [reply scss/input/_layout.scss:209]: I could not find any references for `.k-fieldset` in the sources of Kendo UI for jQuery or Angular. Do we use it anywhere I'm missing or should we erase it?

File: scss/input/_layout.scss (hunk 6 above)
Comment (inikolova) [reply scss/input/_layout.scss]: .k-alert-error has font-size defined too, so we cannot merge these rules

File: scss/input/_layout.scss (hunk 14 above)
Comment (gyoshev) [reply scss/input/_layout.scss]: Good catch, it seems that we can use `span:not(.k-widget)`. Not too good performance, though :disappointed: ...

File: scss/input/_layout.scss (hunk 12 above)
Comment (inikolova) [reply scss/input/_layout.scss:209]: I was not able to find any references to .k-fieldset, too.

File: scss/input/_layout.scss (hunk 5 above)
Comment (inikolova) [reply scss/input/_layout.scss]: This class is specific to some sections in a form, so I prefer keeping the selector as it is

File: scss/input/_layout.scss (hunk 7 above)
Comment (inikolova) [reply scss/input/_layout.scss]: Fixed

File: scss/input/_layout.scss (hunk 13 above)
Comment (inikolova) [reply scss/input/_layout.scss:286]: No, I made the needed changes

File: scss/input/_layout.scss (hunk 16 above)
Comment (inikolova) [reply scss/input/_layout.scss]: Fixed, as @gyoshev suggested

File: scss/input/_layout.scss (hunk 15 above)
Comment (inikolova) [reply scss/input/_layout.scss]: This rule was modified as **display: flex** was added

File: scss/input/_layout.scss (hunk 8 above)
Comment (inikolova) [reply scss/input/_layout.scss]: Fixed

File: scss/input/_layout.scss (hunk 3 above)
Comment (inikolova) [reply scss/input/_layout.scss]: Fixed

**Pull request merged**